### ISSUE TITLE
SwitchBot Curtain Support

### ISSUE THREAD

**Issue by DJay-X** [OpenWonderLabs/homebridge-switchbot-openapi#6]
@donavanbecker First of all I want to say thank you for starting this project and for your time and work sharing with the community. 👍 
I can't wait to get my token from SwitchBot Support to start testing.

Are you planing to integrate all SwitchBot devices like in my case using Bot, Curtain, Humidifier?
Of course I'm happy to help testing. 😇

Thanks again.

**Comment by donavanbecker** [comment]
@DJay-X, I haven't got my token yet either. I can't guarantee the plugin will work out of the box once I get my token, since I haven't been able to test it with it yet.

Once I get my Token and I can get humidifier supported then I will start working on other SwitchBot devices.

**Comment by jahnkenneth** [comment]
Curtains, please. 😀 And I will buy you a beer! 
If you post a paypal address, I am sure the donations will come!

**Comment by donavanbecker** [comment]
@jahnkenneth I will have to wait for my own Token code before we can get to excited. As long as I can get my SwitchBot stuff into the plugin, then I will start with others. I believe there are a few plugins currently that work with the curtains.

**Comment by DJay-X** [comment]
@donavanbecker @whatUwant 
With v1.0.2-beta.2 the curtain bots respond. But I can't even describe how. 😅 It's a bit strange.

They don't respond instantly and not to the % level I choose. I.e. I turn to 100% (open) the App is showing 72% and the curtain is somewhere else. 

I'm also not sure how I can help. 🤓 It's not refreshing and not responding like expected. Not smooth at all. At this beta level not reliable. But the plugin is communicating and the log is looking good I think.

```
[27.12.2020, 13:44:23] [SwitchBot] Discovered Living Room Curtain left Curtain
[27.12.2020, 13:44:23] [SwitchBot] Adding new accessory: Living Room Curtain left Curtain
[27.12.2020, 13:44:23] [SwitchBot] Discovered Living Room Curtain right Curtain
[27.12.2020, 13:44:23] [SwitchBot] Adding new accessory: Living Room Curtain right Curtain
[27.12.2020, 13:55:20] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,0, commandType: command
[27.12.2020, 13:55:32] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,0, commandType: command
[27.12.2020, 13:55:33] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,0, commandType: command
[27.12.2020, 13:57:13] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,28, commandType: command
[27.12.2020, 13:57:13] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,50, commandType: command
[27.12.2020, 13:57:13] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,52, commandType: command
[27.12.2020, 13:57:13] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,52, commandType: command
[27.12.2020, 13:58:24] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,0, commandType: command
[27.12.2020, 13:58:25] [SwitchBot] Sending request to SwitchBot API. command: setPosition, parameter: 0,ff,0, commandType: command
```

**Comment by donavanbecker** [comment]
@DJay-X until I get my curtain bots, I we will have to keep these in beta for now.

**Comment by DJay-X** [comment]
I would expect @SwitchBot-Wonderlabs is sending you all Bots free and with Express delivery. 😌

If there is anything I can do to help ongoing testing, please of course let me know. 

Thanks again to all who contribute to bring the plugin forward step by step.

**Comment by whatUwant** [comment]
The API is not reporting correct status of the device (see https://github.com/OpenWonderLabs/SwitchBotAPI/issues/4). Before that's resolved, I guess we'll just have to wait.
The fact that the API is on the cloud instead of locally on the hub also makes the response not as quick as hoped.

**Comment by SwitchBot-Wonderlabs** [comment]
I have modified the judgment condition of PostionState,  the state updates will be correct normally.
But, if Hubmini control fails, the status request will continue. Needs to be fixed.

Another issue is that the initialization state is not correct in Home APP.
![homeapp](https://user-images.githubusercontent.com/73269232/103194537-f565d100-491a-11eb-97bb-8ec7d3736cd1.png)

**Comment by DJay-X** [comment]
@donavanbecker @SwitchBot-Wonderlabs 

Crash of entire homebridge because of SwitchBot API / Token Problem?
Not depending on v1.0.1 or v1.0.2-beta.3.

If something like this happens, probably the Plugin shouldn't crash the entire Homebridge in the future.
[28.12.2020, 14:42:24] Got SIGTERM, shutting down Homebridge...

I hope the attached log helps, to get everything up and running again fast.

```
Error: Request failed with status code 401
    at createError (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/adapters/http.js:260:11)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
 {
  config: {
    url: 'https://api.switch-bot.com/v1.0/devices',
    method: 'get',
    headers: {
      Accept: 'application/json, text/plain, */*',
      Authorization: 'xxxxxxxxx',
      'Content-Type': 'application/json; charset=utf8',
      'User-Agent': 'axios/0.21.1'
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    adapter: [Function: httpAdapter],
    responseType: 'json',
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    validateStatus: [Function: validateStatus],
    data: undefined
  },
  request: <ref *1> ClientRequest {
    _events: [Object: null prototype] {
      socket: [Function (anonymous)],
      abort: [Function (anonymous)],
      aborted: [Function (anonymous)],
      connect: [Function (anonymous)],
      error: [Function (anonymous)],
      timeout: [Function (anonymous)],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 7,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    destroyed: false,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    _defaultKeepAlive: true,
    useChunkedEncodingByDefault: false,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: 0,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: false,
      _SNICallback: null,
      servername: 'api.switch-bot.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'api.switch-bot.com',
      _readableState: [ReadableState],
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular *1],
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(async_id_symbol)]: 56,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object],
      [Symbol(RequestTimeout)]: undefined
    },
    _header: 'GET /v1.0/devices HTTP/1.1\r\n' +
      'Accept: application/json, text/plain, */*\r\n' +
      'Authorization: xxxxxxxxx\r\n' +
      'Content-Type: application/json; charset=utf8\r\n' +
      'User-Agent: axios/0.21.1\r\n' +
      'Host: api.switch-bot.com\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _keepAliveTimeout: 0,
    _onPendingData: [Function: noopPendingOutput],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: [Object],
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      scheduling: 'fifo',
      maxTotalSockets: Infinity,
      totalSocketCount: 5,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'GET',
    maxHeaderSize: undefined,
    insecureHTTPParser: undefined,
    path: '/v1.0/devices',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 401,
      statusMessage: 'Unauthorized',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular *1],
      responseUrl: 'https://api.switch-bot.com/v1.0/devices',
      redirects: [],
      [Symbol(kCapture)]: false,
      [Symbol(RequestTimeout)]: undefined
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    host: 'api.switch-bot.com',
    protocol: 'https:',
    _redirectable: Writable {
      _writableState: [WritableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _options: [Object],
      _ended: true,
      _ending: true,
      _redirectCount: 0,
      _redirects: [],
      _requestBodyLength: 0,
      _requestBodyBuffers: [],
      _onNativeResponse: [Function (anonymous)],
      _currentRequest: [Circular *1],
      _currentUrl: 'https://api.switch-bot.com/v1.0/devices',
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      authorization: [Array],
      'content-type': [Array],
      'user-agent': [Array],
      host: [Array]
    }
  },
  response: {
    status: 401,
    statusText: 'Unauthorized',
    headers: {
      date: 'Mon, 28 Dec 2020 14:13:27 GMT',
      'content-type': 'application/json',
      'content-length': '26',
      connection: 'close',
      'x-amzn-requestid': 'xxxxxxxxx',
      'x-amzn-errortype': 'UnauthorizedException',
      'x-amz-apigw-id': 'xxxxxxxxx='
    },
    config: {
      url: 'https://api.switch-bot.com/v1.0/devices',
      method: 'get',
      headers: [Object],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      adapter: [Function: httpAdapter],
      responseType: 'json',
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      maxBodyLength: -1,
      validateStatus: [Function: validateStatus],
      data: undefined
    },
    request: <ref *1> ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      socket: [TLSSocket],
      _header: 'GET /v1.0/devices HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Authorization: xxxxxxxxx\r\n' +
        'Content-Type: application/json; charset=utf8\r\n' +
        'User-Agent: axios/0.21.1\r\n' +
        'Host: api.switch-bot.com\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: noopPendingOutput],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/v1.0/devices',
      _ended: true,
      res: [IncomingMessage],
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'api.switch-bot.com',
      protocol: 'https:',
      _redirectable: [Writable],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    data: { message: 'Unauthorized' }
  },
  isAxiosError: true,
  toJSON: [Function: toJSON]
}
Error: Request failed with status code 401
    at createError (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/usr/local/lib/node_modules/homebridge-switchbot-openapi/node_modules/axios/lib/adapters/http.js:260:11)
    at IncomingMessage.emit (events.js:327:22)
    at endReadableNT (internal/streams/readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
 {
  config: {
    url: 'https://api.switch-bot.com/v1.0/devices',
    method: 'get',
    headers: {
      Accept: 'application/json, text/plain, */*',
      Authorization: 'xxxxxxxxx',
      'Content-Type': 'application/json; charset=utf8',
      'User-Agent': 'axios/0.21.1'
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    adapter: [Function: httpAdapter],
    responseType: 'json',
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    validateStatus: [Function: validateStatus],
    data: undefined
  },
  request: <ref *1> ClientRequest {
    _events: [Object: null prototype] {
      socket: [Function (anonymous)],
      abort: [Function (anonymous)],
      aborted: [Function (anonymous)],
      connect: [Function (anonymous)],
      error: [Function (anonymous)],
      timeout: [Function (anonymous)],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 7,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    destroyed: false,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    _defaultKeepAlive: true,
    useChunkedEncodingByDefault: false,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: 0,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: false,
      _SNICallback: null,
      servername: 'api.switch-bot.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'api.switch-bot.com',
      _readableState: [ReadableState],
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular *1],
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(async_id_symbol)]: 56,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object],
      [Symbol(RequestTimeout)]: undefined
    },
    _header: 'GET /v1.0/devices HTTP/1.1\r\n' +
      'Accept: application/json, text/plain, */*\r\n' +
      'Authorization: xxxxxxxxx\r\n' +
      'Content-Type: application/json; charset=utf8\r\n' +
      'User-Agent: axios/0.21.1\r\n' +
      'Host: api.switch-bot.com\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _keepAliveTimeout: 0,
    _onPendingData: [Function: noopPendingOutput],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: [Object],
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      scheduling: 'fifo',
      maxTotalSockets: Infinity,
      totalSocketCount: 5,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'GET',
    maxHeaderSize: undefined,
    insecureHTTPParser: undefined,
    path: '/v1.0/devices',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 401,
      statusMessage: 'Unauthorized',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular *1],
      responseUrl: 'https://api.switch-bot.com/v1.0/devices',
      redirects: [],
      [Symbol(kCapture)]: false,
      [Symbol(RequestTimeout)]: undefined
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    host: 'api.switch-bot.com',
    protocol: 'https:',
    _redirectable: Writable {
      _writableState: [WritableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _options: [Object],
      _ended: true,
      _ending: true,
      _redirectCount: 0,
      _redirects: [],
      _requestBodyLength: 0,
      _requestBodyBuffers: [],
      _onNativeResponse: [Function (anonymous)],
      _currentRequest: [Circular *1],
      _currentUrl: 'https://api.switch-bot.com/v1.0/devices',
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      authorization: [Array],
      'content-type': [Array],
      'user-agent': [Array],
      host: [Array]
    }
  },
  response: {
    status: 401,
    statusText: 'Unauthorized',
    headers: {
      date: 'Mon, 28 Dec 2020 14:13:27 GMT',
      'content-type': 'application/json',
      'content-length': '26',
      connection: 'close',
      'x-amzn-requestid': 'xxxxxxxxx',
      'x-amzn-errortype': 'UnauthorizedException',
      'x-amz-apigw-id': 'xxxxxxxxx='
    },
    config: {
      url: 'https://api.switch-bot.com/v1.0/devices',
      method: 'get',
      headers: [Object],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      adapter: [Function: httpAdapter],
      responseType: 'json',
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      maxBodyLength: -1,
      validateStatus: [Function: validateStatus],
      data: undefined
    },
    request: <ref *1> ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      socket: [TLSSocket],
      _header: 'GET /v1.0/devices HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Authorization: xxxxxxxxx\r\n' +
        'Content-Type: application/json; charset=utf8\r\n' +
        'User-Agent: axios/0.21.1\r\n' +
        'Host: api.switch-bot.com\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: noopPendingOutput],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/v1.0/devices',
      _ended: true,
      res: [IncomingMessage],
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'api.switch-bot.com',
      protocol: 'https:',
      _redirectable: [Writable],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    data: { message: 'Unauthorized' }
  },
  isAxiosError: true,
  toJSON: [Function: toJSON]
}
```

**Comment by donavanbecker** [comment]
> @donavanbecker @SwitchBot-Wonderlabs
> 
> Crash of entire homebridge because of SwitchBot API / Token Problem?
> Not depending on v1.0.1 or v1.0.2-beta.3.
> 
> If something like this happens, probably the Plugin shouldn't crash the entire Homebridge in the future.
> [28.12.2020, 14:42:24] Got SIGTERM, shutting down Homebridge...
> 
> I hope the attached log helps, to get everything up and running again fast.
> 

@DJay-X can you please post this in a new issue?

**Comment by matclothier** [comment]
This is working for me but some of my Curtains only show as 99% closed in the SwitchBot app even though they are fully closed. I tried recalibrating but that doesn’t help. As they are only 99% closed they get stuck on the status closing in Homebridge/Apple home app. Can I be changed so that 99% closed is treated as fully closed?

**Comment by donavanbecker** [comment]
@matclothier, I have added an option in the config to allow you to set the min and max. Since I don't have curtains yet I can test this out. Can you install the latest beta? set your config using the new option as you can see below and report back if it is working correctly?

<img width="780" alt="Screen Shot 2020-12-28 at 3 07 27 PM" src="https://user-images.githubusercontent.com/9875439/103243128-68bd1200-491e-11eb-893c-5c64a8db7488.png">

**Comment by matclothier** [comment]
I updated and set the option but it still shows the Curtains that are at 99% as closing. Any logging I can send or anything to help?

**Comment by matclothier** [comment]
I just noticed it is failing to get the status since the update with a 401 error in the log

**Comment by donavanbecker** [comment]
> I updated and set the option but it still shows the Curtains that are at 99% as closing. Any logging I can send or anything to help?

okay we I have pushed another beta. can you try that?

**Comment by donavanbecker** [comment]
> I just noticed it is failing to get the status since the update with a 401 error in the log

```401 | Unauthorized | Authorization for the API is required, but the request has not been authenticated.```

can you post logs?

**Comment by matclothier** [comment]
Here is the main log from homebridge. Let me know if you need something different. #
[homebridge.log.txt](https://github.com/OpenWonderLabs/homebridge-switchbot-openapi/files/5748928/homebridge.log.txt)

**Comment by donavanbecker** [comment]
> Here is the main log from homebridge. Let me know if you need something different. #
> [homebridge.log.txt](https://github.com/OpenWonderLabs/homebridge-switchbot-openapi/files/5748928/homebridge.log.txt)

can you open a new issue? with this

**Comment by SwitchBot-Wonderlabs** [comment]
Please test the v1.0.2-beta.8.

**Comment by matclothier** [comment]
Fixed. Great job

**Comment by donavanbecker** [comment]
> Fixed. Great job

That's Awesome. Is the config option to allow you to set the min and max, also working for you?

**Comment by matclothier** [comment]
The closing works great now but when open they show as 99% open

**Comment by jahnkenneth** [comment]
Could you post this to nom so I can install via Hoops please? :-)

**Comment by donavanbecker** [comment]
> Could you post this to nom so I can install via Hoobs please? :-)

I would recommend switch to homebridge, since we develop this plugin around homebridge not HOOBs.

**Comment by jahnkenneth** [comment]
Even if I use HomeBridge Ui X I am not able to install betas if they are not published to NPM

**Comment by donavanbecker** [comment]
> Even if I use HomeBridge Ui X I am not able to install betas if they are not published to NPM

You are able too. You install the latest and then you select other versions and select beta. As you see below.
![image](https://user-images.githubusercontent.com/9875439/103311087-02e48f00-49df-11eb-9f28-3d81cd631a54.jpeg)![image](https://user-images.githubusercontent.com/9875439/103311092-06781600-49df-11eb-9c7d-d4dea12b3b0c.jpeg)

**Comment by matclothier** [comment]
I have installed beta 10 and for each pair of Curtains the main Curtain still shows as 99% open but the second Curtain now shows as closed and doesn’t respond. I have them configured via SwitchBot as pairs of Curtains

**Comment by donavanbecker** [comment]
@matclothier can you clear cache on those curtains? and then see what it's like in the home app? My curtain should be here in about a week and a half. Once I get them, I will be able to actually troubleshoot locally.

**Comment by matclothier** [comment]
Sorry that is the same. I’ll wait until you get your to test. Thanks for everything.

Thanks,

Mat
+44(0)7990 591055
________________________________

**Comment by jahnkenneth** [comment]
The plugin has been working prefect for the last few days, but today I get this error:
[31/12/2020, 12:27:42] [SwitchBot] Failed to Discover Devices. "Request failed with status code 401"

**Comment by donavanbecker** [comment]
`v1.1.0` releases initial support for curtains. I will be working into improve this support overtime.
